%% LyX 2.1.0beta2 created this file.  For more info, see http://www.lyx.org/.
%% Do not edit unless you really know what you are doing.
\documentclass[english]{article}
\usepackage{mathpazo}
\renewcommand{\sfdefault}{lmss}
\renewcommand{\ttdefault}{lmtt}
\usepackage[T1]{fontenc}
\usepackage[latin9]{inputenc}
\usepackage{geometry}
\geometry{verbose,tmargin=2.5cm,bmargin=2.5cm,lmargin=2.5cm,rmargin=2.5cm}
\usepackage{color}
\usepackage{babel}
\usepackage{amssymb}
\usepackage[numbers]{natbib}
\usepackage[unicode=true,pdfusetitle,
 bookmarks=true,bookmarksnumbered=false,bookmarksopen=false,
 breaklinks=false,pdfborder={0 0 0},backref=false,colorlinks=true]
 {hyperref}
\hypersetup{
 citecolor=magenta, urlcolor=blue}
\usepackage{breakurl}

\makeatletter
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% User specified LaTeX commands.
%\VignetteIndexEntry{Hot-spring proteins in CHNOSZ}
%\VignetteEngine{knitr::knitr}

% so DOIs in bibliography show up as hyperlinks
\newcommand*{\doi}[1]{\href{http://dx.doi.org/#1}{doi: #1}}

\makeatother

\begin{document}
<<setup, include=FALSE, cache=FALSE>>=
library(knitr)
## set global chunk options
opts_chunk$set(fig.path='figure/hotspring-', cache.path='cache/hotspring-', fig.align='center', fig.show='hold', par=TRUE)
## set code/output width to be 80
options(width=80)
## tune details of base graphics (http://yihui.name/knitr/hooks)
knit_hooks$set(par=function(before, options, envir){
if (before && options$fig.show!='none') par(mar=c(4,4,1,1),cex.lab=.95,cex.axis=.9,mgp=c(2,.7,0),tcl=-.3)
})
@


\title{Hot-spring proteins in CHNOSZ}

\maketitle
This is based on R code in the Supporting Information for the papers.
All code used is shown in the blocks, but much of the text output
(particularly in the figure-generating code) has been suppressed.
In order to keep the code here as short and efficient as possible,
some aspects of the published figures including manual labeling and
higher resolution are not all reproduced; these changes are noted
where possible. Also, some tricks (e.g. using ``xtabs'' to create
the BLAST frequency table, and ``revisit'' to calculate the DGtr)
have been introduced in order to streamline the code.

Load CHNOSZ and the thermodynamic database. The thermodynamic data
for the methionine sidechain group used in the paper has been superseded
in CHNOSZ\_0.9-9 \citep{LD12}, but the older dataset is still available
using \texttt{add.obigt()}:

<<libraryCHNOSZ>>=
library(CHNOSZ)
data(thermo)
add.obigt()
@

First the shared constants

<<shared>>=
# names of the sampling locations
sites <- c("N", "S", "R", "Q", "P")
sitenames <- paste("bison", sites, sep="")
# the measured temperatures (degrees C) and pHs
bison.T <- c(93.3, 79.4, 67.5, 65.3, 57.1)
bison.pH <- c(7.350, 7.678, 7.933, 7.995, 8.257)
@

Load the proteins

<<proteins>>=
# read the amino acid compositions
aa.annot <- read.aa(system.file("extdata/protein/DS11.csv", package="CHNOSZ"))
aa.phyla <- read.aa(system.file("extdata/protein/DS13.csv", package="CHNOSZ"))
@

Setup names, colors, line types

<<names>>=
# the annotation-derived classifications
classes <- unique(aa.annot$protein)
# the names of the phyla in alphabetical order (Deinococcus-Thermus at end)
phyla.abc <- sort(unique(aa.phyla$organism))[c(1:7,9:11,8)]
# an abbreviation for Dein.-Thermus
phyla.abbrv <- phyla.abc
phyla.abbrv[[11]] <- "Dein.-Thermus"
# colors used to identify the phyla in the barchart, based on Wu and Eisen, 2008 with modifications
phyla.cols <- c("#f48ba5", "#f2692f", "#cfdd2a",
  "#962272", "#87c540", "#66c3a2", "#12a64a", "#f58656",
  "#ee3237", "#25b7d5", "#3953a4")
# line types used to identify the phyla
phyla.lty <- c(1:6, 1:5)
@

Function to setup the basis species

<<setup.basis>>=
# function to load basis species
setup.basis <- function() {
  basis(c("HCO3-", "H2O", "NH3", "HS-", "H2", "H+"))
  basis(c("HCO3-", "NH3", "HS-", "H+"), c(-3, -4, -7, -7.933))
}
@

Function to calculate $\log a_{\mathrm{H_{2}}}$, according to Equation
(2) in \citep{DS13}.

<<logaH2>>=
get.logaH2 <- function(T) -11 + T * 3/40
@

Set the temperature limits (\texttt{Tlim}) over which to perform the
calculations.

<<Tlim>>=
Tlim <- c(50, 100)
@


\section{Average oxidation state of carbon}

The average oxidation state of carbon ($\overline{Z}_{\mathrm{C}}$)
is calculated using
\begin{equation}
\overline{Z}_{\mathrm{C}}=\frac{Z-n_{\mathrm{H}}+2\left(n_{\mathrm{O}}+n_{\mathrm{S}}\right)+3n_{\mathrm{N}}}{n_{\mathrm{C}}}\,,\label{eq:ZC}
\end{equation}
where $n_{\mathrm{C}}$, $n_{\mathrm{H}}$, $n_{\mathrm{N}}$, $n_{\mathrm{O}}$
and $n_{\mathrm{S}}$ are the numbers of the indicated subscripted
elements in the chemical formula of a protein or other chemical species,
and $Z$ is the net charge of the chemical species.



<<ZCplot, fig.width=5, fig.height=5, out.width='.49\\textwidth'>>=
# 2011 plot
plot(0, 0, xlim=c(-0.5, 5), ylim=c(-0.27, -0.11), xlab="location", xaxt="n", ylab=expression(bar(italic(Z))[C]))
axis(1, at=1:5)
col <- c("green", rep("black", 20))
lwd <- c(3, rep(1, 20))
clab <- c("hydrolase", "overall", "protease", "oxidoreductase", "transport", "membrane", "permease")
pf.annot <- protein.formula(aa.annot)
ZC.annot <- ZC(pf.annot)
for(i in 1:length(classes)) {
  lines(1:5, ZC.annot[(1:5)+5*(i-1)], col=col[i], lwd=lwd[i])
  if(classes[i] %in% clab) text(0.8, ZC.annot[1+5*(i-1)], classes[i], adj=1)
}
title(main="annotations")

# 2013 plot
pf.phyla <- protein.formula(aa.phyla)
ZC.phyla <- ZC(pf.phyla)
# set up plot
plot(0, 0, xlim=c(1, 5), ylim=c(-0.27, -0.11), xlab="location", ylab=as.expression(quote(bar(italic(Z))[C])))
for(i in 1:length(phyla.abc)) {
  # which of the model proteins correspond to this phylum
  iphy <- which(aa.phyla$organism==phyla.abc[i])
  # the locations (of 1, 2, 3, 4, 5) where this phylum is found
  ilocs <- match(aa.phyla$protein[iphy], sitenames)
  # the plotting symbol: determined by alphabetical position of the phylum
  points(ilocs, ZC.phyla[iphy], pch=i-1, cex=1.2)
  # a line to connect same phyla occurring at adjacent sites
  inlocs <- rep(NA, 5)
  inlocs[ilocs] <- ilocs
  lines(inlocs, ZC.phyla[iphy][match(1:5, ilocs)])
}
legend("bottomright", pch=0:10, legend=phyla.abbrv, bg="white", cex=0.9)
title(main="major phyla")
@

DOI links for these plots as they appeared in the publications: \href{http://dx.doi.org/10.1371/journal.pone.0022782.g004}{plot 1},
\href{http://dx.doi.org/10.1371/journal.pone.0072395.g001}{plot 2}.
Compared to the papers, a different \emph{y}-axis scale is used here,
in order to have same scale on both plots; permease at lower $\overline{Z}_{\mathrm{C}}$
does not appear in plot 1 here.


\section{Comparison old and new {[}Met{]}}

Make some $T-\log a_{\mathrm{H_{2}}}$ protein metastable equilibrium
predominance diagrams using old and new methionine sidechain {[}Met{]}
parameters.

<<methionine, fig.width=6, fig.height=4, message=FALSE, results='hide', cache=TRUE>>=
par(mfrow=c(2, 3))
for(j in 1:2) {
  # use old [Met] for first row and new [Met] for second row
  if(j==1) add.obigt(force=TRUE) else data(thermo)
  # setup basis species and proteins
  setup.basis()
  ip.annot <- add.protein(aa.annot)
  # make the plots
  for(annot in c("overall", "transferase", "synthase")) {
    ip <- ip.annot[aa.annot$protein==annot]
    a <- affinity(T=c(50, 100), H2=c(-7, -4), iprotein=ip)
    diagram(a, fill=NULL, names=1:5, normalize=TRUE)
    # add logaH2-T line
    lines(par("usr")[1:2], get.logaH2(par("usr")[1:2]), lty=3)
    # add a title
    title(main=annot)
  }
}
@

DOI link to original figure: \href{http://10.1371/journal.pone.0072395.g002}{link}.
The original was made with higher resolution (256) and some labels
were manually placed.


\section{Relative abundance calculation}

Function to return the fractional abundances based on BLAST counts,
stored in the 'ref' column of aa.phyla.

<<alpha.blast>>=
alpha.blast <- function() {
  out <- xtabs(ref ~ protein + organism, aa.phyla)
  # put it in correct order
  out <- out[c(1,5:2), c(1:7,9:11,8)]
  # turn counts into fractions (sum=1)
  out <- out/rowSums(out)
  return(as.data.frame(out))
}
@

Function to calculate metastable equilibrium degrees of formation
of proteins (normalized to residues) as a function of $\log a_{\mathrm{H_{2}}}$
for a specified location. Note: in the call to affinity, change 101
to 1001 to reproduce calculations in paper.

<<alpha.equil>>=
alpha.equil <- function(i=1) {
  # which rows of the table correspond to this location
  iloc <- which(aa.phyla$protein==sitenames[i])
  # order the names and counts to go with the alphabetical phylum list
  iloc <- iloc[order(match(aa.phyla$organism[iloc], phyla.abc))]
  # set up basis species, with pH specific for this location
  setup.basis()
  basis("pH", bison.pH[i])
  # calculate affinities of reactions to form the proteins over the logaH2 range
  a <- affinity(H2=c(-11, -1, 101), T=bison.T[i], iprotein=ip.phyla[iloc])
  # calculate the metastable equilibrium activities of the residues
  e <- equilibrate(a, loga.balance=0, normalize=TRUE, stay.normal=TRUE)
  # remove the logarithms to get relative abundances (sum=1 because loga.balance=0)
  a.residue <- 10^sapply(e$loga.equil, c)
  colnames(a.residue) <- aa.phyla$organism[iloc]
  # the BLAST profile
  a.blast <- alpha.blast()
  # calculate Gibbs energy of transformation (DGtr) and find optimal logaH2
  iblast <- match(colnames(a.residue), colnames(a.blast))
  r <- revisit(e, "DGtr", log10(a.blast[i, iblast]), plot.it=FALSE)
  # return the calculated activities, logaH2 range, DGtr values, and optimal logaH2
  return(list(alpha=a.residue, H2vals=a$vals[[1]], DGtr=r$H, logaH2.opt=r$xopt))
}
@

Now we're ready to make a plot! We start by adding the proteins with
amino acid composition that were read from the file. Then plot the
degrees of formation as a function of $\log a_{\mathrm{H_{2}}}$ at
sites 1, 3 and 5, but record the optimal values of $\log a_{\mathrm{H_{2}}}$
at all 5 sites.

<<alphaplot, fig.width=8, fig.height=6, tidy=FALSE, message=FALSE, results='hide', cache=TRUE>>=
ip.phyla <- add.protein(aa.phyla)
layout(matrix(1:6, ncol=3), heights=c(2, 1))
# keep track of the optimal logaH2
logaH2.opt <- numeric()
for(i in 1:5) {
  # get the equilibrium degrees of formation and the optimal logaH2
  ae <- alpha.equil(i)
  logaH2.opt <- c(logaH2.opt, ae$logaH2.opt)
  if(i %in% c(1, 3, 5)) {
    iphy <- match(colnames(ae$alpha), phyla.abc)
    # top row: equilibrium degrees of formation
    thermo.plot.new(xlim=range(ae$H2vals), ylim=c(0, 0.5), xlab=axis.label("H2"),
      ylab=expression(alpha[equil]), yline=2, cex.axis=1, mgp=c(1.8, 0.3, 0))
    for(j in 1:ncol(ae$alpha)) {
      lines(ae$H2vals, ae$alpha[, j], lty=phyla.lty[iphy[j]])
      ix <- seq(1, length(ae$H2vals), length.out=11)
      ix <- head(tail(ix, -1), -1)
      points(ae$H2vals[ix], ae$alpha[, j][ix], pch=iphy[j]-1)
    } 
    title(main=paste("site", i))
    legend("topleft", pch=iphy-1, lty=phyla.lty[iphy], legend=phyla.abbrv[iphy], bg="white")
    # bottom row: Gibbs energy of transformation and position of minimum
    thermo.plot.new(xlim=range(ae$H2vals), ylim=c(0, 1/log(10)), xlab=axis.label("H2"),
      ylab=expr.property("DGtr/2.303RT"), yline=2, cex.axis=1, mgp=c(1.8, 0.3, 0))
    lines(ae$H2vals, ae$DGtr)
    abline(v=ae$logaH2.opt, lty=2)
    abline(v=get.logaH2(bison.T[i]), lty=3, lwd=1.5)
    if(i==1) legend("bottomleft", lty=c(3, 2), lwd=c(1.5, 1), bg="white",
      legend=c("Equation 2", "optimal"))
  }
}
@

\href{http://dx.doi.org/10.1371/journal.pone.0072395.g003}{DOI link}
to published version of the figure.


\section{Activity of hydrogen comparison}

Now that we have computed the activities of hydrogen that minimize
the Gibbs energy of transformation between the metastable equilibrium
of model proteins and observed phylum abundances, let's plot them
and also compare them with the field measurements of varioius redox
proxies and with the linear equation used to fit the spatial distribution
of proteins from the 2011 study \citep{DS11} (reprinted as Equation
2 in 2013 \citep{DS13}):
\begin{equation}
\log a_{\mathrm{H_{2_{\left(aq\right)}}}}=-11+3/40\times T\mathrm{\left(^{\circ}C\right)}\label{eq:logaH2}
\end{equation}



\subsection{Conversion of field measurements}


\paragraph{ORP}

Oxidation-reduction potential (ORP), measured in the field using a
portable probe and pH/voltmeter, can be converted to Eh by adding
the potential of the reference electrode, in this case silver-silver
chloride (Ag/AgCl) in saturated KCl. Lacking the high-temperature
potentials of this electrode, the following equation from Ref. \citep{BPJ85}
for Ag/AgCl (1M KCl) is used, with temperature in degrees Celsius
and potential in volts.

<<E.AgAgCl>>=
E.AgAgCl <- function(T) {
  0.23737 - 5.3783e-4 * T - 2.3728e-6 * T^2 - 2.2671e-9 * (T+273)
}
@

Together with ORP, temperature and pH will all be needed. Data from
Bison Pool and another hot spring shown in Fig. 9 of Ref. \citep{DS11}
are both included, but for simplicity here they are lumped together.
The ORP values have units of mV.

<<meters>>=
T.ORP <- c(93.9, 87.7, 75.7, 70.1, 66.4, 66.2)
pH.ORP <- c(8.28, 8.31, 7.82, 7.96, 8.76, 8.06)
ORP <- c(-258, -227, -55, -58, -98, -41)
@

Convert ORP to effective values of $\log a_{\mathrm{H_{2}}_{\left(aq\right)}}$.
First convert to Eh (in volts). Then convert Eh to pe. Then combine
Eh and pH with the law of mass action for
\[
2e^{-}+2\mathrm{H^{+}}\rightleftharpoons\mathrm{H_{2}}_{\left(aq\right)}
\]
to calculate $\log a_{\mathrm{H_{2}}_{\left(aq\right)}}$. The law
of mass action is the condition where the activity quotient ($Q$)
of the reaction is equal to the equilibrium constant ($K$).

<<ORP2Eh>>=
Eh <- ORP/1000 + E.AgAgCl(T.ORP)
pe <- convert(Eh, "pe", T=convert(T.ORP, "K"))
logK.ORP <- subcrt(c("e-", "H+", "H2"), c(-2, -2, 1), c("aq", "aq", "aq"), T=T.ORP)$out$logK
logaH2.ORP <- logK.ORP - 2*pe - 2*pH.ORP
@


\paragraph{Sulfide/Sulfate}

For sulfide/sulfate, assign activities that are equal to concentrations
(in molal units) measured in the field season of 2005, when the biofilm
samples were collected for metagenomic sequencing, and use the law
of mass action for
\[
\mathrm{HS^{-}}+4\mathrm{H_{2}O}\rightleftharpoons\mathrm{SO_{4}}^{-2}+\mathrm{H^{+}}+4\mathrm{H_{2}}_{\left(aq\right)}
\]
to calculate an effective activity of hydrogen. Sulfide was determined
spectrophotometrically at the hot spring, and sulfate was determined
using ion chromatography on water samples returned from the field.

<<sulfur>>=
loga.HS <- log10(c(4.77e-6, 2.03e-6, 3.12e-7, 4.68e-7, 2.18e-7))
loga.SO4 <- log10(c(2.10e-4, 2.03e-4, 1.98e-4, 2.01e-4, 1.89e-4))
logK.S <- subcrt(c("HS-", "H2O", "SO4-2", "H+", "H2"), c(-1, -4, 1, 1, 4), 
  state=c("aq", "liq", "aq", "aq", "aq"), T=bison.T)$out$logK
logaH2.S <- (logK.S + bison.pH - loga.SO4 + loga.HS) / 4
@


\paragraph{Dissolved Oxygen}

Calculate the effective activity of hydrogen corresponding to the
dissolved oxygen concentration using the law of mass action for
\[
0.5\mathrm{O_{2}}_{\left(aq\right)}+\mathrm{H_{2}}_{\left(aq\right)}\rightleftharpoons\mathrm{H_{2}O}\,.
\]
Convert the dissolved oxygen concentrations, in mg/L, to molarities
(mol/L) and use these values to set the activity of $\mathrm{O_{2}}_{\left(aq\right)}$.

<<oxygen>>=
DO <- c(0.173, 0.776, 0.9, 1.6, 2.8)
logaO2 <- log10(DO/1000/32)
logK <- subcrt(c("O2", "H2", "H2O"), c(-0.5, -1, 1), c("aq", "aq", "liq"), T=bison.T)$out$logK
logaH2.O <- 0 - 0.5*logaO2 - logK
@


\paragraph{Plot it!}

The points in the left plot represent proxies for redox potential
based on ORP, sulfur and oxygen measurements, while the points in
in right plot are the optimal values of $\log a_{\mathrm{H_{2}}}$
for the microbial phylum abundance distribution. Notably, the various
redox indicators are not in equilibrium with each other (the $\log a_{\mathrm{H_{2}}}$
values are different), but they all show an increase in apparent $\log a_{\mathrm{H_{2}}}$
with temperature. The dotted line in both plots portrays Equation
(\ref{eq:logaH2}).

<<logaH2plot, fig.width=4, fig.height=4, out.width='.49\\textwidth', cache=TRUE>>=
# 2011 plot
plot(Tlim, get.logaH2(Tlim), xlim=Tlim, ylim=c(-45,0),
  xlab=axis.label("T"), ylab=axis.label("H2"), type="l", lty=3)
points(T.ORP, logaH2.ORP, pch=15)
lines(T.ORP, logaH2.ORP, lty=2)
points(bison.T, logaH2.O, pch=16)
lines(bison.T, logaH2.O, lty=2)
points(bison.T, logaH2.S, pch=17)
lines(bison.T, logaH2.S, lty=2)
llab <- c("Equation 2", "ORP", "dissolved oxygen", "sulfate/sulfide")
text(c(65, 80, 80, 74), c(-4, -25, -40, -11), llab)

# 2013 plot
plot(Tlim, get.logaH2(Tlim), xlim=Tlim, ylim=c(-11,-2),
  xlab=axis.label("T"), ylab=axis.label("H2"), type="l", lty=3)
lines(bison.T, logaH2.opt, lty=2)
points(bison.T, logaH2.opt, pch=21, bg="white")
text(90, -5.3, "Equation 2")
text(66, -9, "optimal parameterization\nfor observed\nphylum abundances", adj=0)
@

DOI links to original versions: \href{http://dx.doi.org/10.1371/journal.pone.0022782.g009}{plot 1},
\href{http://dx.doi.org/10.1371/journal.pone.0072395.g004}{plot 2}.


\section{Relative abundance comparison}

\bibliographystyle{unsrtnat}
\bibliography{vig}

\end{document}
