---
title: "An Introduction to [CHNOSZ](http://www.chnosz.net)"
author: "Jeffrey M. Dick"
date: "`r Sys.Date()`"
output:
  tufte::tufte_html:
    tufte_features: ["background"]
    css: "vig.css"
    toc: true
  tufte::tufte_handout:
    citation_package: natbib
    latex_engine: xelatex
  tufte::tufte_book:
    citation_package: natbib
    latex_engine: xelatex
vignette: >
  %\VignetteIndexEntry{An Introduction to CHNOSZ}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
bibliography: vig.bib
link-citations: yes
---

```{r options, include=FALSE}
options(width = 80)
options(digits = 6)
```

```{r setup, include=FALSE}
# invalidate cache when the tufte version changes
knitr::opts_chunk$set(tidy = FALSE, cache.extra = packageVersion('tufte'))
options(htmltools.dir.version = FALSE)
knitr::knit_hooks$set(small.mar = function(before, options, envir) {
    if (before) par(mar = c(4, 4, .1, .1))  # smaller margin on top and right
})
```

# About

This document introduces the basic functionality of CHNOSZ, a package for the [R software environment](http://r-project.org).
For more information on R, see "An Introduction to R" and other documents in [The R Manuals](http://cran.r-project.org/manuals.html).

Since 2006, CHNOSZ has been developed to support a research project on the thermodynamic properties of proteins.
Over time, the package has grown to calculate thermodynamic properties of reactions and to construct equilibrium chemical activity diagrams for both inorganic and organic systems.
Development since 2009 has focused on the calculation of chemical affinities and metastable equilibrium for large numbers of proteins with applications to -omic (metagenomic and proteomic) data in various environments.
Nevertheless, the database and functions are flexible, allowing one to model the relative stabilities of proteins, minerals or aqueous species using similar commands.

## Installing and loading CHNOSZ

If you have just installed R for the first time, and you have an internet connection, installing CHNOSZ should be as simple as selecting the "Install packages from CRAN" or similar menu item in the R GUI or using the following command to start the package installation process:^[Or, install the package from a local package file, which you can download from [CRAN](http://cran.r-project.org) or from the [CHNOSZ website](http://chnosz.net).
][]()
```{r install_CHNOSZ, eval=FALSE}
install.packages("CHNOSZ")
```

Then load the CHNOSZ package to make its functions available in your working session.
```{r library_CHNOSZ}
library(CHNOSZ)
```

Then load the thermo object, which contains the thermodynamic database and is also where your system settings will be stored.
```{r data_thermo}
data(thermo)
```

## Organization of major functions

CHNOSZ is made up of a set of functions and supporting datasets.
The major components of the package are shown in the figure below, which is an updated version of the flowchart from @Dic08 (boxes – functions; ellipses – datasets; bold text – major user functions).
![CHNOSZ flowchart](CHNOSZ.png)

Many functions in CHNOSZ have no side effects.
That is, the function only returns a result; to use the result in other functions, it can be assigned to a variable with `<-`.
Major functions without side effects in CHNOSZ are:

* `info()`: search for species in the thermodynamic database
* `subcrt()`: calculate the thermodynamic properties of species and reactions
* `affinity()`: calculate the affinities of formation reactions using given chemical activities
* `equilibrate()`: calculate the equilibrium chemical activities of the species of interest
* `diagram()`: plot the results

Some functions in CHNOSZ do have side effects: they modify the `thermo` data object in the current R session.
In the text (not code) of this document, the names of these functions are shown in <span style="color:red">red</span>.
The major functions with side effects are:

* <span style="color:red">`basis()`</span>: set the basis species and their chemical activities
* <span style="color:red">`species()`</span>: set the species of interest and their (non-equilibrium) chemical activities
* <span style="color:red">`data(thermo)`</span>: reset the database, restoring all settings to their default values.

The following pseudocode shows a common sequence of commands.
In actual usage, the ... are replaced by arguments that define the chemical makeup and range of conditions of the system:
```{r, eval=FALSE}
basis(...)
species(...)
a <- affinity(...)
e <- equilibrate(a)  ## optional
diagram(e)           ## or diagram(a)
data(thermo)         ## clear system settings
```

Some experimental functions are available:

* (**experimental**) using `revisit()` to calculate/plot summary statistics of the chemical activities of the species of interest and `findit()` to search for combinations of activities of basis species, temperature and/or pressure that optimize those statistics.

# Thermodynamic database and chemical formulas

While an attempt has been made to provide a primary database (`OBIGT.csv`) that is generally internally consistent, all thermodynamic data, calculations, and examples are provided *as is*.^[
For crucial problems, check not only the accuracy of the database, but also the *suitability of the data* for your problem.
If there is any doubt about the suitability of data, please consult the primary sources (which can be located using `browse.refs()`; see Section XXX).
As with the data, please check the *suitability and accuracy of the available computational techniques*, at least by comparing the construction and output of the examples to the primary sources cited in each help page.
Some examples without references demonstrate the more experimental features of CHNOSZ. 
]
Where possible, data with known or suspected inconsistencies have been placed into a secondary database (`OBIGT-2.csv`) that should be regarded as experimental.

## The `info()` function

The `info()` function provides an interface to the thermodynamic database packaged with CHNOSZ.
Suppose you are interested in the thermodynamic properties aqueous ethylene.
You can search for the species by name:
```{r info_ethylene}
info("ethylene")
```

Multiple entries exist for ethylene; the index of the `aq` (aqueous) species is returned by default.
A second argument can be used to specify a different physical state:
```{r info_ethylene_gas}
info("ethylene", "gas")
```

Knowing that aqueous ethylene is species number 88 in the database, you can again use `info()` to retrieve the set of standard molal thermodynamic properties and equations of state parameters:
```{r info_88}
info(88)
```

This number can be used as an argument (`ispecies`) for other functions in CHNOSZ to uniquely identify any species; some commonly used functions also accept the species names.
Liquid water is species number 1; it has NA entries in the database because specialized functions are used to compute its properties:
```{r info_info_water}
info(info("water"))
```

## Fuzzy searches

Calling `info()` with a string that does not exactly match the name of any species invokes a fuzzy search of the database:
```{r width180, include=FALSE}
options(width = 180)
```
```{r info_acid}
info("acid")
```
```{r width80, include=FALSE}
options(width = 80)
```

The message includes e.g. "uracil" and "metacinnabar" because their names have some similarity to the search term.

As "ribose" is the name of a species in the database, to find species with similar names, add an extra character to the search:
```{r info_ribose}
info(" ribose")
```

The messages may be useful for browsing the database, but owing to their ambiguous results, these fuzzy searches return an `NA` value for the species index.

## Counting elements, chemical formulas, `ZC()`

Continuing with the example of ethylene, let's look at its chemical formula:
```{r info_88_formula}
info(88)$formula
```

We can use the `makeup()` function to count the elements in the formula, followed by `as.chemical.formula()` to return to the formula:
```{r makeup_88}
makeup(88)
as.chemical.formula(makeup(88))
```

For organic species, a simple calculation of the average oxidation state of carbon ($Z_C$) is possible given the species index, chemical formula, or elemental count:
```{r ZC_88}
ZC(88)
ZC(info(88)$formula)
ZC(makeup(88))
```

# Calculating thermodynamic properties

To calculate the standard molal properties of species and reactions, use `subcrt()`.^[
The inspiration for this function (as well as its name), and the source of the FORTRAN subroutine used to calculate the thermodynamic properties of H<sub>2</sub>O, is SUPCRT (Johnson et al. 1992).]<sup> [-@JOH92]</sup>
If no reaction coefficients are given, `subcrt()` calculates the standard molal properties of invididual species:
```{r subcrt_water}
subcrt("water")
```

That uses the default temperature and pressure settings: equally-spaced temperature grid from 0 to 350 °C at _P_<sub>sat</sub>.^[_P_<sub>sat</sub> is 1 bar below 100 °C, or the pressure of liquid-vapor coexistence at higher temperatures.]
The columns in the output are^[The corresponding units are °C (`T`), bar (`P`), g cm<sup>-3</sup> (`rho`), cal mol<sup>-1</sup> (`G` and `H`), cal K<sup>-1</sup> mol<sup>-1</sup> (`S`), cm<sup>3</sup> mol<sup>-1</sup> (`V`), and cal K<sup>-1</sup> mol<sup>-1</sup> (`Cp`).] temperature, pressure, density of water, logarithm of the equilibrium constant (only meaningful for reactions; see below), standard molal Gibbs energy and enthalpy of formation from the elements, standard molal entropy, volume, and heat capacity.

A custom temperature-pressure grid can be specified.
Here, we calculate the properties of H<sub>2</sub>O on a _T_,_P_ grid in the supercritical region, with conditions grouped by pressure:
```{r subcrt_water_grid}
subcrt("water", T=c(400, 500, 600), P=c(200, 400, 600), grid="P")$out$water
```

```{r subcrt_water_plot, fig.margin=TRUE, fig.width=4, fig.height=4, small.mar=TRUE, dpi=50, out.width="100%", echo=FALSE, message=FALSE, fig.cap="Isothermal contours of density (g cm<sup>-3</sup>) and pressure (bar) of H<sub>2</sub>O.", cache=TRUE}
H2O <- subcrt("water", T=seq(0, 1000, 100), P=c(NA, seq(1, 500, 1)), grid="T")
H2O <- H2O$out$water
plot(H2O$P, H2O$rho, type="l")
```
The additional operations (`$out$water`) are used to extract a specific part of the results; this can be used with e.g. `write.table()` or `plot()` for further processing:
```{r subcrt_water_plot, eval=FALSE}
```

## Changing units

The default units of temperature, pressure, and energy are °C, bar, and calories.
The functions <span style="color:red">`T.units()`</span>, <span style="color:red">`P.units()`</span>, and <span style="color:red">`E.units()`</span> can be used to change the units used by various functions in CHNOSZ.
What is the Gibbs energy (J/mol) of aqueous methane at 298.15 K and 0.1 MPa?
```{r methane_units, message=FALSE}
T.units("K")
P.units("MPa")
E.units("J")
subcrt("methane", T=298.15, P=0.1)$out$methane$G
data(thermo)  ## restore default settings
```

A related function, `convert()`, can be used to convert given values between units.
Let's convert the standard Gibbs energy of aqueous methane listed in the database from cal/mol to J/mol:
```{r methane_G, message=FALSE}
convert(info(info("methane"))$G, "J")
```

As expected, we get the same result from both operations.

# Properties of reactions

## Reaction definitions

To calculate the thermodynamic properties of reactions, give the names of species, the physical states (optional), and reaction coefficients as the arguments to `subcrt()`.
Here we calculate properties for the dissolution of CO<sub>2</sub>:
```{r CO2_dissolution}
subcrt(c("CO2", "CO2"), c("gas", "aq"), c(-1, 1))
```

<p> <!--- needed to make the citation appear correction 20170202 -->
In order to make a plot like that shown by Manning et al. [-@MSS13, Figure 18], let's run more calculations and store the results.
In addition to the reaction definition, we specify a greater number of temperature points than the default:
```{r dissolution, echo=FALSE, message=FALSE}
T <- seq(0, 350, 10)
CO2 <- subcrt(c("CO2", "CO2"), c("gas", "aq"), c(-1, 1), T=T)$out$logK
CO <- subcrt(c("CO", "CO"), c("gas", "aq"), c(-1, 1), T=T)$out$logK
CH4 <- subcrt(c("CH4", "CH4"), c("gas", "aq"), c(-1, 1), T=T)$out$logK
logK <- data.frame(T, CO2, CO, CH4)
```
```{r dissolution_plot, fig.margin=TRUE, fig.width=4, fig.height=4, small.mar=TRUE, dpi=50, out.width="100%", echo=FALSE, message=FALSE, fig.cap="Equilibrium constants calculated for dissolution of CO<sub>2</sub>, CO, and CH<sub>4</sub>.", cache=TRUE}
matplot(logK[, 1], logK[, -1], type="l",
        xlab=axis.label("T"), ylab=axis.label("logK"))
```
```{r dissolution, eval=FALSE}
```

Now we can make the plot.
Here, the `axis.label()` function of CHNOSZ is used to create formatted axis labels:
```{r dissolution_plot, eval=FALSE}
```

## Unbalanced reactions

## Setting the basis species

## Auto-balancing reactions

# Basic diagrams

## Setting the species of interest

## Thermodynamic potential: `affinity()`

## Predominance: the `diagram()` function

## Labels, colors, etc.

# Equilibration

## Getting from affinity to equilibrium

## Setting the constraints

# Other things you can do with affinity

## Buffers

## T-, P-, activity-transect

# Other things you can do with equilibrate

## Choosing different balancing constraints

# Other things you can do with diagrams

## Mosaic diagrams

## Groups of species

# Proteins

## Group additivity

## Sources of amino acid data

## Ionization

## Normalizing for different lengths
